### ISSUE TITLE
variable shadowing bug in loop not caught by compiler and crashing at runtime.

### ISSUE THREAD

**Issue by IwateKyle** [rust-lang/rust#86565]
After many months of using Rust - I was going back thru "The Rust Programming Language (book)" implementing the guessing game.  

I tried this code:

```rust
// .....
fn main() {
 10     println!("Guess the numer!");
  9
  8     let secret_number = rand::thread_rng().gen_range(1,101);
  7     println!("The secret number is: {}", secret_number);
  6
  5     println!("Please input your guess.");
  4
  3     let mut guess = String::new();
  2
  1     loop {
16          io::stdin().read_line(&mut guess).expect("Failed to read line");
  1         let guess:u32 = guess.trim().parse().expect("Please type a number!");
  2
  3
  4         println!("You guessed: {}", guess);
  5
  6         match guess.cmp(&secret_number) {
  7             Ordering::Less => println!("Too small!"),
  8             Ordering::Greater => println!("Too big!"),
  9             Ordering::Equal => {println!("You win!"); break;},
 10         }
 11     }
 12 }
 // ....
```
There is a bug in the above code which after finding it I thought Rust would have / should have caught?  Correct me if I'm wrong.

The program compiles fine with no warnings or errors.  When I run the program - I answer the the question to deliberately get it wrong - I am informed that the answer is wrong.

Then the guess variable is shadowed in the loop, the program compiles fine, then crashes at runtime after guessing a number a SECOND time.

If I circumvent the loop ( NO LOOP) as in the code below by just calling two iterations worth of read_line() then the program properly fails at compile time.
```rust
//    loop {
    1         io::stdin().read_line(&mut guess).expect("Failed to read line");
    2         let guess:u32 = guess.trim().parse().expect("Please type a number!");
    3
    4
    5
    6         println!("You guessed: {}", guess);
    7
    8         match guess.cmp(&secret_number) {
    9             Ordering::Less => println!("Too small!"),
   10             Ordering::Greater => println!("Too big!"),
   11             Ordering::Equal => {println!("You win!"); },
   12         }
>> 13         io::stdin().read_line(&mut guess).expect("Failed to read line");
>> 14         let guess:u32 = guess.trim().parse().expect("Please type a number!");
   15 //    }
```

The problem: when in a loop - it passes compilation - 
   I expect that it should NOT pass compilation and should raise the same compilation error as if there were NO loop.

If I'm wrong - I apologize ahead of time.

**Comment by mjguynn** [comment]
According to the documentation for `read_line`, it appends to the existing string rather than replacing/clearing it. So if you typed in `foo` for the first input and `5` for the second input, the string would be `foo5` on the second iteration, causing the program to panic when `expect` is called.

**Comment by fee1-dead** [comment]
This is not a bug in the compiler. You *can* shadow variables.
```rs
let mut guess = String::new();
loop {
    io::stdin().read_line(&mut guess).expect("Failed to read line");
    // ^ `guess` as a variable of `u32` is not visible to this line...
    let guess:u32 = guess.trim().parse().expect("Please type a number!");
    // because it is only defined here in the loop...
}
```

**Comment by IwateKyle** [comment]
> This is not a bug in the compiler. You _can_ shadow variables

Hi fee1-dead, thanks for your thoughts.  I think *maybe* you didn't understand my post.  Your code introduces the same problem. And before responding, I actually copied your code, compiled it and ran it.  So the problem still exists.

I think you might have misunderstood the point about being able to shadow variables in a loop.  Its not that you *can't*, it's that you *can* and the compiler isn't catching it, therefore causing a crash.  :)  Hope that helps.

**Comment by fee1-dead** [comment]
No, when you shadow variables, they will not be in the same location. The compiler knows that they are different and independent of each other. For your problem, read [this comment](https://github.com/rust-lang/rust/issues/86565#issuecomment-866460090). There isn't any bug in the compiler that caused your program to crash. There is a bug in your code that caused it.

**Comment by IwateKyle** [comment]
I feel like I stated that fairly clearly.... when I said ```There is a bug in the above code .... ```  So, I'm not sure how to proceed with our discourse.  And, the *very same* code that *you* game me crashes.  In other words, your code has the very same bug.  I don't want to imply that you didn't read what I said, so we'll just chalk it up to "I didn't explain the problem very well."  Thank you for your input @fee1-dead

**Comment by IwateKyle** [comment]
@fee1-dead in a second attempt - I edited the problem description above.  I hope this helps.

**Comment by fee1-dead** [comment]
Okay. But this issue is still invalid. The compiler should not catch the bug in your code, because the compiler doesn't know about what you are trying to achieve.

 1. Name shadowing is okay. Your bug is not caused by name shadowing.
 
Well, but why did it error in your second example?

```rs
let guess: u32 = guess.trim().parse().expect("Please type a number!");
// ^ After this line, the previous variable with type `String` is now inaccessible.
...
io::stdin().read_line(&mut guess).expect("Failed to read line");
// ^ read_line expects `&mut String`, while you are passing a `&mut u32`.
let guess: u32 = guess.trim().parse().expect("Please type a number!");
// ^ There is no `.trim()` for `u32`.
```

That is pretty obvious.

 2. I did not give you code where I think it would behave correctly. I directly used your example and cleaned up line numbers and unnecessary details.

 3. In your loop, you call `read_line` on the same string. You can't even make a lint for the "bug", because it can be a valid use case, to reuse the string passed to read_line.

 4. The compiler isn't catching your "bug" for your second example. Read what the compiler says.

 5. This is more suitable for Stack Overflow / Users forum.

 6. You seem to be a bit angry, so please calm down...

**Comment by jonas-schievink** [comment]
This is working as intended, so closing

**Comment by IwateKyle** [comment]
@fee1-dead 
Anger No.  
Disappointment - a big dose of it.

I took a chance on posting here based on something that I *still* feel has merit  - even after reading everything you have put forth and Jonas's detailed comment.  I'm mostly disappointed that because I, myself along with many others have a great deal of desire to see the success of Rust multiplied.  Often people regard programming technologies by the level of community they accompany.  I didn't sense any of that in this dialogue.  I sensed someone who just wants to say that someone is wrong without really acknowledging what is being said.  

I've never made claims that I want Rust to catch my logic bugs.  I know where the bug is.  This is why I'm bringing it out in the open to show that the compiler catches the same code error in once case, but not the other the the only thing removed was 

loop {

}

I have gathered a great deal of important information today and it has been a great lesson to me.   May the journey continue.

**Comment by fee1-dead** [comment]
> Often people regard programming technologies by the level of community they accompany. I didn't sense any of that in this dialogue. I sensed someone who just wants to say that someone is wrong without really acknowledging what is being said.

Unfortunately, I do not represent this community. But I would only agree with realistic and helpful suggestions. Think about this, if no one responded to your issue and someone just added a tag named "invalid" and closed your issue. Wouldn't that be a pretty bad community? I think not necessarily.

If you opened this issue hoping someone would fix it, and end up thinking the same way even after the issue is closed, that should tell you something is wrong. This reveals something about yourself, and I hope you would stop responding and take a minute to read all the comments so far. I came here to try to help you, and not to simply prove you wrong. What do I get from proving that you are wrong anyways? No, I don't feel good after proving that someone is wrong, just like I don't feel bad after admitting that I was wrong.

### Real talk

Take a step back from "level of community and disappointment". I will try to help you understand this, again. Not to prove that you are wrong. First, let me address your claim that you only removed "loop { }". Look at these examples that you gave:

```rs
// first example
loop {
    io::stdin().read_line(&mut guess).expect("Failed to read line");
    let guess:u32 = guess.trim().parse().expect("Please type a number!");
    match guess.cmp(&secret_number) {
        Ordering::Less => println!("Too small!"),
        Ordering::Greater => println!("Too big!"),
        Ordering::Equal => {println!("You win!"); break; }
    }
}
```

```rs
// second example
// loop {
    io::stdin().read_line(&mut guess).expect("Failed to read line");
    let guess:u32 = guess.trim().parse().expect("Please type a number!");
    println!("You guessed: {}", guess);
    match guess.cmp(&secret_number) {
        Ordering::Less => println!("Too small!"),
        Ordering::Greater => println!("Too big!"),
        Ordering::Equal => {println!("You win!"); },
    }
    io::stdin().read_line(&mut guess).expect("Failed to read line");
    let guess:u32 = guess.trim().parse().expect("Please type a number!");
// }
```

Well, looks like you just commented out `loop {` and `}` and there is an error, whoops.

Secondly, I already told you that the error that you are encountering after removing the loop and adding two lines is not "catching your bug". It is reporting a type error. Well, it looks like you are following chapter 2, but if you followed the instructions correctly, you would initialize the string inside the loop. (It also looks like it includes a lot of `--snip--` that can make people confused about where the code actually end up, probably a good issue for the book)

> We create a variable named guess. But wait, doesn’t the program already have a variable named guess? It does, but Rust allows us to shadow the previous value of guess with a new one. This feature is often used in situations in which you want to convert a value from one type to another type. Shadowing lets us reuse the guess variable name rather than forcing us to create two unique variables, such as guess_str and guess for example. (Chapter 3 covers shadowing in more detail.)

Chapter 3 talks about this exactly:

> The other difference between mut and shadowing is that because we’re effectively creating a new variable when we use the let keyword again, we can change the type of the value but reuse the same name. For example, say our program asks a user to show how many spaces they want between some text by inputting space characters, but we really want to store that input as a number:

You will also learn about scopes in [chapter 4](https://doc.rust-lang.org/book/ch04-01-what-is-ownership.html#variable-scope).
Basically, after the previous iteration of the loop, `guess: u32` is now out of scope and dropped before you parse another `guess: u32` from user input. This is the exact reason for `guess: String`. It is outside the `loop {}` therefore reused by all iterations of the loop.

In conclusion, there can be no error that correctly identifies your logic error, and it only identified a type error.

As an exercise for learning more about Rust, consider this example, why does the compiler not catch my "logic error"? All I did was adding curly brackets.

```rs
fn main() {
    // --snip--
    let mut guess = String::new();
    io::stdin().read_line(&mut guess).expect("Failed to read line");
    {
        let guess:u32 = guess.trim().parse().expect("Please type a number!");
        println!("You guessed: {}", guess);
        match guess.cmp(&secret_number) {
            Ordering::Less => println!("Too small!"),
            Ordering::Greater => println!("Too big!"),
            Ordering::Equal => {println!("You win!"); },
        }
    }
    io::stdin().read_line(&mut guess).expect("Failed to read line");
    let guess:u32 = guess.trim().parse().expect("Please type a number!");
}
```

**Comment by IwateKyle** [comment]
@fee1-dead 
Thank you for your genuine and lengthy, and clear response.  After the closing of our conversation yesterday - I took your advice and basically reposted the above question [on Stackoverflow here.](https://stackoverflow.com/questions/68097792/variable-shadowing-bug-in-loop-not-caught-by-compiler-and-crashing-at-runtime).

I finally understand my extreme lacking understanding of shadowing does and doesn't work - but also about read_line().  This was truly a lack of understanding on my part.  

As far as our discussion yesterday - granted maybe the wrong place to have it - I was hoping for an answer similar to- 

__"Aha - this is very common struggling point - I can understand why you'd see that these two same sets of code appear to be doing the same thing whether they be inside a loop or unrolled.  But in fact they are doing different things. "__ then maybe point me to a link where others have had the same problem.  

I would have probably moved on at that point and closed my own ticket.  I didn't feel that anyone was seeing the problem as I was seeing it - so I didn't get that sense of confirmation that I was maybe still correct - and just poorly describing the problem.   It appears to be a sticking point for others as I learned in the stack overflow conversation.

In the end, I'm grateful for the time you invested yesterday and today helping out someone across the world to expand their understanding of these issues one small grain at a time.  Next time - I'll start off on the Stackoverflow side before posting here.
